File: 3.py
from common import read_file
from collections import namedtuple
import re


def parse(filename):
    lines = read_file(filename)
    xs = len(lines[0])
    ys = len(lines)
    symbols = []
    gears = []
    items = []
    item = namedtuple('Item', ['value', 'x1', 'x2', 'y'])
    for y in range(ys):
        num_matches = re.finditer(r'(\d+)', lines[y])
        for num in num_matches:
            x1, x2 = num.span()
            items.append(
                item(
                    value=int(num.group()),
                    x1=x1,
                    x2=x2,
                    y=y
                )
            )
        for x in range(xs):
            char = lines[y][x]
            if char != '.' and not char.isdigit():
                if char == '*':
                    gears.append(item('*', x, x+1, y))
                symbols.append((x, y))
    return symbols, gears, items, (xs, ys)


def get_surrounding_coords(item, size_of_table):
    max_x, max_y = size_of_table
    border = []
    # top and bottom
    for j in [item.y-1, item.y+1]:
        for i in range(item.x1 - 1, item.x2 + 1):
            if (0 <= i <= max_x - 1) and (0 <= j <= max_y - 1):
                border.append((i, j))
    # left side
    if 0 <= item.x1 - 1:
        border.append((item.x1 - 1, item.y))
    # right side
    if item.x1 <= max_x - 1:
        border.append((item.x2, item.y))
    return border


def get_sum_of_items(items, symbols, size_of_table):
    sum_of_items = 0
    for item in items:
        surr = get_surrounding_coords(item, size_of_table)
        for sym in symbols:
            if sym in surr:
                sum_of_items += item.value
    return sum_of_items


def get_gear_ratios(items, gears, size_of_table):
    gear_connections = [get_surrounding_coords(g, size_of_table) for g in gears]
    connections = set()
    # id's for gears
    for i, g in enumerate(gear_connections):
        for item in items:
            for x in range(item.x1, item.x2):
                if (x, item.y) in g:
                    connections.add((i, item.value))
    s2 = 0
    d = {}
    for c in connections:
        if c[0] in d.keys():
            s2 += c[1] * d[c[0]]
        else:
            d[c[0]] = c[1]
    return s2


if __name__ == '__main__':
    symbol_coords, gear_coords, item_coords, size = parse('input/input3.txt')

    print(get_sum_of_items(item_coords, symbol_coords, size))
    print(get_gear_ratios(item_coords, gear_coords, size))
    # 507214
    # 72553319

File: julia/3.jl
### A Pluto.jl notebook ###
# v0.19.36

using Markdown
using InteractiveUtils

# ╔═╡ 6812b28d-3fce-4dfd-a7ce-0dae037c1c6d
struct Item
	value::Int
	range::UnitRange
	y::Int
end

# ╔═╡ 5deef660-a99f-11ee-04ff-879927efe650
example = split(raw"""467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..""")

# ╔═╡ dd63b7b4-6efd-45c0-8ddb-399d5afffb89
function get_surrounding_coords(item::Item, sizeoftable)
	max_x, max_y = sizeoftable
	border = Tuple{Int, Int}[]
	for j in [item.y - 1, item.y + 1]
		for i in first(item.range)-1:last(item.range)+1
			if (1 <= i <= max_x) && (1 <= j <= max_y)
				push!(border, (i, j))
			end
		end
	end
	if 1 < first(item.range)
		push!(border, (first(item.range) - 1, item.y))
	end
	if first(item.range) <= max_x
		push!(border, (last(item.range) + 1, item.y))
	end
	return border
end

# ╔═╡ 62863571-4b70-400d-ad83-539d1d156100
function parsed(lines)
	rc = length(example)
	cc = length(example[1])
	symbols = Tuple{Int, Int}[]
	gears = Item[]
	items = Item[]
	for row in 1:rc
		num_matches = findall(r"(\d+)", lines[row])
		for m in num_matches
			push!(items, Item(parse(Int, lines[row][m]), m, row))
		end
		for column in 1:cc
			char = lines[row][column]
			if char != '.' && !isdigit(char)
				if char == '*'
					push!(gears, Item(0, column:column+1, row))
				end
				push!(symbols, (column, row))
			end
		end
	end
	return items, gears, symbols, (cc, rc)
end

# ╔═╡ 01fc9b01-b2ef-4ff6-a296-4ac73de9b68d
items, gears, symbols, sizeoftable = parsed(example)

# ╔═╡ 216569ca-1573-435c-b758-d554a0cae40c
function get_sum_of_items(items, symbols, sizeoftable)
	sum_of_items = 0
	for item in items
		surr = get_surrounding_coords(item, sizeoftable)
		for sym in symbols
			if sym in surr
				sum_of_items += item.value
			end
		end
	end
	return sum_of_items
end

# ╔═╡ d6cbbb24-40f9-4c74-b6ef-7d09c5517ba9
get_sum_of_items(items, symbols, sizeoftable)

# ╔═╡ Cell order:
# ╠═6812b28d-3fce-4dfd-a7ce-0dae037c1c6d
# ╠═5deef660-a99f-11ee-04ff-879927efe650
# ╠═dd63b7b4-6efd-45c0-8ddb-399d5afffb89
# ╠═62863571-4b70-400d-ad83-539d1d156100
# ╠═01fc9b01-b2ef-4ff6-a296-4ac73de9b68d
# ╠═216569ca-1573-435c-b758-d554a0cae40c
# ╠═d6cbbb24-40f9-4c74-b6ef-7d09c5517ba9
